src/ReplaySession.cc

```diff
@@ -1257,6 +1257,40 @@ Completion ReplaySession::flush_syscallbuf(ReplayTask* t,
   return COMPLETE;
 }
 
+Completion ReplaySession::patch_vsyscall(ReplayTask* t, const StepConstraints& constraints)
+{
+  TicksRequest ticks_request;
+  if (!compute_ticks_request(t, constraints, &ticks_request)) {
+    return INCOMPLETE;
+  }
+
+  remote_code_ptr vsyscall_entry = current_trace_frame().regs().ip();
+  bool added = t->vm()->add_breakpoint(vsyscall_entry, BKPT_INTERNAL);
+  ASSERT(t, added);
+  auto complete = continue_or_step(t, constraints, ticks_request, RESUME_CONT);
+  t->vm()->remove_breakpoint(vsyscall_entry, BKPT_INTERNAL);
```

Comment:
Can `patch_next_syscall` use this approach too, so we can merge the two code paths?

Reply:
Seems like if we did that, we wouldn't need the extra trace state.

Reply:
For that, I'm a bit worried about JITs introducing syscall instructions in a way that hides them from our breakpoints. I suppose that's a concern here too, but I'm less worried about it, since nobody should be emitting these anymore, and I'm not aware of any existing code that would jit such jumps.

Reply:
Yeah OK that makes sense.